chain/client/src/chunks_delay_tracker.rs

```diff
@@ -0,0 +1,188 @@
+use std::collections::{BTreeMap, HashMap};
+use std::time::Instant;
+
+use near_primitives::types::{BlockHeight, ShardId};
+
+use crate::metrics;
+use std::cmp::max;
+
+/// Provides monitoring information about the delays of receiving blocks and their corresponding chunks.
+/// Keeps timestamps of a limited number of blocks before the current head.
+/// Keeps timestamps of all blocks and chunks from the future.
+/// Doesn't make any assumptions about shard layout or the number of shards. If the head was moved forward, we assume that all requirements were satisfied.
+/// This object provides two metrics:
+/// 1) max delay between discovering a block and receiving the last chunk from that block.
+/// 2) number of blocks with missing chunks.
+/// If a chunk or a block is received multiple times, only the first time is recorded.
+#[derive(Debug, Default)]
+pub(crate) struct ChunksDelayTracker {
+    heights: BTreeMap<BlockHeight, HeightInfo>,
+    // Used as an optimization to delete old blocks only when the head is updated.
+    prev_head_height: BlockHeight,
+}
+
+const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
+
+#[derive(Debug, Default)]
+struct HeightInfo {
+    block_received: Option<Instant>,
+    chunks_received: HashMap<ShardId, Instant>,
+}
+
+impl ChunksDelayTracker {
+    // Blocks below this heights are not tracked.
+    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
+        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    }
+
+    fn remove_old_entries(&mut self, head_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> = self
+            .heights
+            .range(..ChunksDelayTracker::lowest_height(head_height))
+            .map(|(&k, _v)| k)
+            .collect();
+        for h in heights_to_drop {
+            self.heights.remove(&h);
+        }
+    }
+
+    fn update_chunks_metric(&mut self, head_height: BlockHeight) {
+        if head_height == self.prev_head_height {
+            return;
+        }
+        let mut max_delay = 0;
+        for (_, v) in self.heights.range(..=head_height) {
```

Comment:
Could you add a comment for what this function tries to do and the definition of this metric?